### PR TITLE
[Breaking] Adds `AMP<F>` dtype

### LINKED ISSUE #424: Add AMP dtype
This depends on #423.

This would let each operation/optimizer choose which precision it uses.

The exact interface needs to be thought out, but here's a potential sketch:

```rust
struct AMP<Lo: Dtype, Hi: Dtype> { ... }

let q: Tensor<Rank1<5>, AMP<f16, f32>, _> = dev.zeros();
```

Things to figure out:
- How would certain layers choose an actual dtype from the two options in Amp? trait implementations?
- Is this even possible? Right now Dtype is just used directly for storage (ie. `Vec<Dtype>`), but this proposal is kind of a Dtype container. Does `Vec<AMP<f16, f32>>` even make sense?
- What are the generics on AMP? Are they even needed (i.e. implied f16/f32)?
- Need to write down exactly what operations are f16/f32, and what tensor/gradients are defaulted to (I think the lower precision).

### PR DESCRIPTION
**Pull request by coreylowman**
Resolves #424 

- [Breaking] Moves Unit/Dtype to new `dtypes` module
- Adds `dfdx::dtypes::AMP<F>`
- impl `Device<AMP<f16>>` for `Cpu`/Cuda`

Ops that cast up to f32 internally:
- sum_to
- adam
- sgd
- rmsprop

Other ops may be auto cast in the future.

### CODE REVIEW

**Comment by nkoppel** [comment]
I'm curious: why `struct Amp<F>(F)` rather than `struct Amp(f16)`? Are there plans to support `Amp<f32>` as well? If not, the generic should be removed in my opinion.

File: src/dtypes/amp.rs [hunk 1]
```diff
@@ -0,0 +1,571 @@
...
+    }
+}
+
+#[cfg(feature = "f16")]
+impl num_traits::AsPrimitive<AMP<half::f16>> for f64 {
+    fn as_(self) -> AMP<half::f16> {
+        AMP(half::f16::from_f64(self))
+    }
+}
+
```
Comment (nkoppel) [inline src/dtypes/amp.rs:187]:
```suggestion
#[cfg(feature = "f16")]
impl<T: num_traits::AsPrimitive<E>, E> num_traits::AsPrimitive<AMP<E>> for T {
    fn as_(self) -> AMP<E> {
        AMP(self.as_())
    }
}
```

File: src/dtypes/amp.rs [hunk 2]
```diff
@@ -0,0 +1,571 @@
...
+#[cfg(feature = "f16")]
+impl crate::shapes::SafeZeros for AMP<half::f16> {}
+
+#[cfg(feature = "f16")]
+impl crate::shapes::Unit for AMP<half::f16> {
+    const ONE: Self = AMP(half::f16::ONE);
+}
+
+#[cfg(feature = "f16")]
+impl crate::shapes::Dtype for AMP<half::f16> {}
```
Comment (nkoppel) [inline src/dtypes/amp.rs:212]:
```suggestion
impl<T: crate::shapes::SafeZeros> crate::shapes::SafeZeros for AMP<T> {}

impl<T:  crate::shapes::Unit> crate::shapes::Unit for AMP<T> {
    const ONE: Self = AMP(T::ONE);
}

impl<T: crate::shapes::Dtype> crate::shapes::Dtype for AMP<T> {}
```

**Comment by coreylowman** [comment]
> I'm curious: why `struct Amp<F>(F)` rather than `struct Amp(f16)`? Are there plans to support `Amp<f32>` as well? If not, the generic should be removed in my opinion.

@nkoppel I was thinking we could have `AMP<bf16>` as well, but I don't actually know if that is done in practice. If not, we can move to what you suggest. Can you check if anyone does that?

**Comment by nkoppel** [comment]
PyTorch supports it, as seen [here](https://pytorch.org/docs/stable/amp.html), so it makes sense to leave room to do that. Might it make sense to use a type alias to refer to `AMP<f16>`  named something like `AMP16` to make this easier to use?

**Comment by coreylowman** [comment]
👍 to type aliases

File: src/dtypes/amp.rs (hunk 1 above)
Comment (coreylowman) [reply src/dtypes/amp.rs]: Actually this leads to unconstrained type error on T

**Pull request merged**